05.wc/wc.rb

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'optparse'
+
+def main
+  options, file_paths = parse_options(ARGV)
+
+  # パイプで接続した場合 $stdin.tty? は false になる
+  if file_paths.empty? && !$stdin.tty?
+    count_stats = count_content($stdin.read)
+    max_width = calculate_max_width([{ stats: count_stats }], nil)
+
+    # file_pathにnilを渡すことで、ファイル名を出力しない
+    output_format(count_stats, options, nil, max_width)
+  else
+    file_info_list, total_stats = collect_file_stats(file_paths)
+    max_width = calculate_max_width(file_info_list, total_stats)
+
+    file_info_list.each do |file_info|
+      output_format(file_info[:stats], options, file_info[:path], max_width)
+    end
+
+    output_format(total_stats, options, 'total', max_width) if file_info_list.size > 1
+  end
+end
+
+def parse_options(argv)
+  opt = OptionParser.new
+  options = {}
+  opt.on('-l') { |v| options[:l] = v }
+  opt.on('-w') { |v| options[:w] = v }
+  opt.on('-c') { |v| options[:c] = v }
+  file_paths = opt.parse(argv)
+  [options, file_paths]
+end
+
+def count_content(content)
+  {
+    lines: content.count("\n"),
+    words: content.split.size,
+    bytes: content.bytesize
+  }
+end
+
+def collect_file_stats(file_paths)
+  file_info_list = []
+  total_stats = { lines: 0, words: 0, bytes: 0 }
+
+  file_paths.each do |file_path|
+    file_stats = read_and_count_file(file_path)
+    next unless file_stats
+
+    file_info_list << { stats: file_stats, path: file_path }
+    total_stats[:lines] += file_stats[:lines]
+    total_stats[:words] += file_stats[:words]
+    total_stats[:bytes] += file_stats[:bytes]
+  end
+
+  [file_info_list, total_stats]
+end
+
+def read_and_count_file(file_path)
+  if !File.exist?(file_path) || File.directory?(file_path)
```

Comment:
```suggestion
  if !File.exist?(file_path) || !File.directory?(file_path)
```

が正なのでは？
でも本当は、

```suggestion
  if !File.exist?(file_path)
```

だけでいいのでは？（公式リファレンスをチェックしてみてください）

Reply:
元々
```ruby
  if !File.exist?(file_path)
    warn "wc: #{file_path}: No such file or directory"
    return
  elsif File.directory?(file_path)
    warn "wc: #{file_path}: Is a directory"
    return
  end
```
ということがやりたかったのですが、なぜか途中で「本家コマンドではどちらの場合も `No such file or directory`のメッセージがでてるぞ！」と勘違いして変な形に直してしまっていました。

実際は、指定したファイルがない場合は`No such file or directory` 、指定したパスがディレクトリの場合は `Is a directory`というエラーメッセージを表示したうえで、その配下のファイルのtotalのカウントを表示する仕様でした。
配下のファイルのtotalのカウントを表示する部分を再現しようと思うとまただいぶロジックが膨らんでしまうので、今回は非対応にして、単に `Is a directory`のエラーメッセージを出すところまでにしました。

https://github.com/cellotak/ruby-practices/pull/12/commits/d14e4b67143f18f488127a7b8634026af0cff4be